### PR TITLE
Add Dockerfile for building containers

### LINKED ISSUE #72: Containerize Setup
* Need to containerize the whole setup
* Can consider hosting in [Github's docker registry](https://github.com/features/packages) itself
* Need to setup automated CI/CD for building docker images with every push
* Perhaps can start from https://github.com/NVAITC/ai-lab as a base image (`FROM` tag in `Dockerfile`, and add zsh, etc for setting up development environment

### PR DESCRIPTION
**Pull request by rajat2004**
Closes https://github.com/rsnk96/Ubuntu-Setup-Scripts/issues/72
Docker Images - https://hub.docker.com/repository/docker/rajat2004/ubuntu-setup-scripts

TODO:
- [x] Change to master branch rather than my fork's branch (Using it for any testing or changes required)
- [x] Create an image and push to Docker Hub, change the link in Readme
- [ ] Travis updates to automatically create an image from the master branch and push? (Probably will timeout)
- [ ] Separate Dockerfiles or maybe just images for only OpenCV, non-CUDA, etc.? (Leaning toward images only)

Maybe, possibly not required or useful
- [ ] Add commands for using this as a non-root user (maybe using `gosu`?)
- [x] Add ZSH
- [ ] Add Anaconda

### CODE REVIEW

**Comment by rajat2004** [comment]
`ubuntu18.04-cuda10.1-cv2-tf` was built after using this image and then running `3-ML-Basic.sh` by hand inside container
Will update the info on the docker images page on the details of the setup, running, etc.

File: README.md [hunk 1]
```diff
@@ -125,10 +126,4 @@ In the description of shortcuts below, if two keys are connected with a `+`, the
 
 
 ## Alternatives
-* A [docker image](https://hub.docker.com/r/varun19299/cvi-iitm/) for this set-up (last updated Jan 30th, 2017)
 * A Ubuntu customization dedicated to [robotics](https://github.com/ahundt/robotics_setup)
-
-
-## To Dos 
-- [x] CI
-- [ ] configuring default wifi settings
```
Comment (rajat2004) [inline README.md:134]: Removed since I don't think we're actually gonna do this or that it could/should be done also

**Comment by rajat2004** [comment]
I think such a minimal Dockerfile should work, and the image with Cuda especially is already bloated
The items left in the list don't seem very useful, or are too time-consuming

Any specific things which should be added?

**Comment by rsnk96** [comment]
> I think such a minimal Dockerfile should work, and the image with Cuda especially is already bloated

Agreed. We could follow some of the steps listed [in this link](https://hackernoon.com/tips-to-reduce-docker-image-sizes-876095da3b34) to minimize the docker image size

> The items left in the list don't seem very useful, or are too time-consuming

I think we should add the other components (zsh/conda/....). If people want a docker container for just using AI libraries, they would go with one of the other 100 options out there. I think where this would become useful is in also setting up the development environment (zsh+zim+conda+aliases+...). What are your thoughts on this?

**Comment by rajat2004** [comment]
> I think we should add the other components (zsh/conda/....). If people want a docker container for just using AI libraries, they would go with one of the other 100 options out there. I think where this would become useful is in also setting up the development environment (zsh+zim+conda+aliases+...). What are your thoughts on this?

I'm still not sure about Anaconda, since that always felt a bit too much to me, but there's no downside as such in creating the images, will look into building them. Zsh is okay, will add that.
Had a thought, after installing OpenCV, we don't really need the entire downloaded folder and the `build` one as well, removing those should reduce by 3-4GB at least
Will try to build the images in the next few days, I specifically kept less information in the Readme so that every time a new image is added, some changes are made, etc, no need to open PR for keeping the info upto date

If the Dockerfile and the Readme updates seem okay, I think that it can be merged, the updated images will be directly added to the Docker Hub

**Comment by rsnk96** [comment]
> I'm still not sure about Anaconda, since that always felt a bit too much to me, but there's no downside as such in creating the images, will look into building them. Zsh is okay, will add that.
Had a thought, after installing OpenCV, we don't really need the entire downloaded folder and the build one as well, removing those should reduce by 3-4GB at least
Will try to build the images in the next few days, I specifically kept less information in the Readme so that every time a new image is added, some changes are made, etc, no need to open PR for keeping the info upto date
If the Dockerfile and the Readme updates seem okay, I think that it can be merged, the updated images will be directly added to the Docker Hub

Agreed, it might be better to keep conda out of it. If someone wants to create another Python environment, they could simply create another container off the docker image

**Comment by rajat2004** [comment]
Tested removing the `opencv` folders after building, reduces the size by 3 GB
Will be updating the images and info on Docker Hub as I create them
This seems good to go to me

Docker image updated to include `zsh`, reduced image size

**Comment by rajat2004** [comment]
Merging, don't think any more updates are required in this PR, if so, can be done in follow-up PRs

**Pull request merged**